[YahooWeather] Thing initialisation error in OpenHab2

YahooWeather not working in the latest snapshot version of OpenHab2, throwing errors as below in openhab.log

In PaperUI, the Thing is showing as

`Status: UNINITIALIZED - HANDLER_REGISTERING_ERROR com.google.common.cache.CacheBuilder.maximumSize(I)Lcom/google/common/cache/CacheBuilder;`

openhab.log extract:
`
2017-01-10 22:25:59.876 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while registering the handler for thing 'yahooweather:weather:23285bad' using factory 'org.eclipse.smarthome.binding.yahooweather.internal.YahooWeatherHandlerFactory@10e71959': java.lang.NoSuchMethodError: com.google.common.cache.CacheBuilder.maximumSize(I)Lcom/google/common/cache/CacheBuilder;
java.util.concurrent.ExecutionException: java.lang.NoSuchMethodError: com.google.common.cache.CacheBuilder.maximumSize(I)Lcom/google/common/cache/CacheBuilder;
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)[:1.8.0_111]
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)[:1.8.0_111]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:188)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:81)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:65)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.registerAndInitializeHandler(ThingManager.java:978)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.thingAdded(ThingManager.java:459)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyTrackers(ThingRegistryImpl.java:218)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutAddedElement(ThingRegistryImpl.java:137)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutAddedElement(ThingRegistryImpl.java:1)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.added(AbstractRegistry.java:123)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:51)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:70)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListenersAboutAddedElement(AbstractProvider.java:74)
	at org.eclipse.smarthome.core.common.registry.AbstractManagedProvider.add(AbstractManagedProvider.java:61)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.add(AbstractRegistry.java:177)
	at org.eclipse.smarthome.io.rest.core.thing.ThingResource.create(ThingResource.java:193)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)[:1.8.0_111]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)[:1.8.0_111]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)[:1.8.0_111]
	at java.lang.reflect.Method.invoke(Method.java:498)[:1.8.0_111]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)[157:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)[157:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)[157:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)[157:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)[157:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)[157:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)[158:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)[155:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)[155:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)[155:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)[155:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)[155:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at com.eclipsesource.jaxrs.publisher.internal.ServletContainerBridge.service(ServletContainerBridge.java:76)[10:com.eclipsesource.jaxrs.publisher:5.3.1.201602281253]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)[81:org.eclipse.jetty.servlet:9.2.19.v20160908]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)[81:org.eclipse.jetty.servlet:9.2.19.v20160908]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71)[172:org.ops4j.pax.web.pax-web-jetty:4.3.0]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)[79:org.eclipse.jetty.security:9.2.19.v20160908]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:287)[172:org.ops4j.pax.web.pax-web-jetty:4.3.0]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)[81:org.eclipse.jetty.servlet:9.2.19.v20160908]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80)[172:org.ops4j.pax.web.pax-web-jetty:4.3.0]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.eclipse.jetty.server.Server.handle(Server.java:499)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)[80:org.eclipse.jetty.server:9.2.19.v20160908]
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)[72:org.eclipse.jetty.io:9.2.19.v20160908]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)[83:org.eclipse.jetty.util:9.2.19.v20160908]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)[83:org.eclipse.jetty.util:9.2.19.v20160908]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_111]
Caused by: java.lang.NoSuchMethodError: com.google.common.cache.CacheBuilder.maximumSize(I)Lcom/google/common/cache/CacheBuilder;
	at org.eclipse.smarthome.binding.yahooweather.handler.YahooWeatherHandler.<init>(YahooWeatherHandler.java:177)
	at org.eclipse.smarthome.binding.yahooweather.internal.YahooWeatherHandlerFactory.createHandler(YahooWeatherHandlerFactory.java:42)
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.registerHandler(BaseThingHandlerFactory.java:98)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.doRegisterHandler(ThingManager.java:593)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.registerHandler(ThingManager.java:572)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.access$29(ThingManager.java:568)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$17.call(ThingManager.java:981)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$17.call(ThingManager.java:1)
	at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:179)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_111]
	... 1 more
`


## Your Environment
* Version used: openhab2 (apt version), snapshot 2.0.0~20170110001819-1
YahooWeather addon version: binding-yahooweather - 2.0.0.SNAPSHOT
* Environment name and version (e.g. Chrome 39, node.js 5.4, Java 8, ...):
Java(TM) SE Runtime Environment (build 1.8.0_111-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.111-b14, mixed mode)
Chrome 55
* Operating System and version: Ubuntu 16.04.1

--- a/extensions/binding/org.eclipse.smarthome.binding.yahooweather/META-INF/MANIFEST.MF
+++ b/extensions/binding/org.eclipse.smarthome.binding.yahooweather/META-INF/MANIFEST.MF
@@ -6,8 +6,7 @@ Bundle-Vendor: Eclipse.org/SmartHome
 Bundle-Version: 0.9.0.qualifier
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-ClassPath: .
-Import-Package: com.google.common.cache,
- com.google.common.collect,
+Import-Package: com.google.common.collect,
  com.google.gson;version="2.2.4",
  org.apache.commons.io,
  org.apache.commons.lang,

--- a/extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java
@@ -16,12 +16,12 @@ import java.net.URL;
 import java.net.URLConnection;
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
 
 import org.apache.commons.io.IOUtils;
 import org.apache.commons.lang.StringUtils;
+import org.eclipse.smarthome.binding.yahooweather.internal.ExpiringCache;
 import org.eclipse.smarthome.config.core.Configuration;
 import org.eclipse.smarthome.config.core.status.ConfigStatusMessage;
 import org.eclipse.smarthome.core.library.types.DecimalType;
@@ -36,10 +36,6 @@ import org.eclipse.smarthome.core.types.State;
 import org.eclipse.smarthome.core.types.UnDefType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import com.google.common.cache.Cache;
-import com.google.common.cache.CacheBuilder;
-import com.google.common.cache.CacheLoader;
 
 /**
  * The {@link YahooWeatherHandler} is responsible for handling commands, which are
@@ -56,6 +52,7 @@ public class YahooWeatherHandler extends ConfigStatusThingHandler {
     private final Logger logger = LoggerFactory.getLogger(YahooWeatherHandler.class);
 
     private final int MAX_DATA_AGE = 3 * 60 * 60 * 1000; // 3h
+    private final int CACHE_EXPIRY = 10 * 1000; // 10s
     private long lastUpdateTime;
 
     private BigDecimal location;
@@ -198,10 +195,10 @@ public class YahooWeatherHandler extends ConfigStatusThingHandler {
         return lastUpdateTime + MAX_DATA_AGE < System.currentTimeMillis();
     }
 
-    private final Cache<String, String> CACHE = CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.SECONDS)
-            .maximumSize(5).build(new CacheLoader<String, String>() {
+    private final ExpiringCache<String, String> CACHE = new ExpiringCache<String, String>(CACHE_EXPIRY,
+            new ExpiringCache.LoadAction<String, String>() {
                 @Override
-                public String load(String query) throws Exception {
+                public String load(String query) throws IOException {
                     try {
                         URL url = new URL("https://query.yahooapis.com/v1/public/yql?format=json&q="
                                 + query.replaceAll(" ", "%20").replaceAll("'", "%27"));
@@ -215,11 +212,7 @@ public class YahooWeatherHandler extends ConfigStatusThingHandler {
             });
 
     private String getWeatherData(String query) throws IOException {
-        try {
-            return CACHE.get(query);
-        } catch (ExecutionException e) {
-            throw new IOException(e.getMessage(), e);
-        }
+        return CACHE.get(query);
     }
 
     private State getHumidity() {

--- a/extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/internal/ExpiringCache.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/internal/ExpiringCache.java
@@ -1,0 +1,66 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.binding.yahooweather.internal;
+
+import java.io.IOException;
+
+/**
+ * This is a simple cache implementation of a single key-value-pair.
+ *
+ * There must be provided an action in order to retrieve/calculate the value for any given key. This action will be
+ * called only if the answer from the last calculation is not valid anymore, i.e. if the key is different or the last
+ * result is expired.
+ *
+ * @author Simon Kaufmann - Initial contribution and API.
+ *
+ * @param <Q> the type of the key
+ * @param <V> the type of the value
+ */
+public class ExpiringCache<Q, V> {
+
+    private final int expiry;
+    private final LoadAction<Q, V> action;
+    private long lastUpdate;
+    private Q lastQuery;
+    private V lastValue;
+
+    public interface LoadAction<Q, V> {
+        V load(Q key) throws IOException;
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param expiry duration in milliseconds for how long a result stays valid
+     * @param action
+     */
+    public ExpiringCache(int expiry, LoadAction<Q, V> action) {
+        this.expiry = expiry;
+        this.action = action;
+    }
+
+    /**
+     * Get the value for the given key - possibly from the cache, if it is still valid.
+     *
+     * @param key the input parameter
+     * @return the value for the given key
+     * @throws IOException
+     */
+    public synchronized V get(Q key) throws IOException {
+        if (lastQuery == null || !lastQuery.equals(key) || isExpired()) {
+            lastValue = action.load(key);
+            lastUpdate = System.currentTimeMillis();
+        }
+        return lastValue;
+    }
+
+    private boolean isExpired() {
+        return lastUpdate + expiry < System.currentTimeMillis();
+    }
+
+}

